### PR TITLE
MODTLR-10: Support for cross-tenant requests

### PR DESCRIPTION
**Pull request by OleksandrVidinieiev**
## Purpose
Add Circulation client able to create requests in a different tenant.

## Approach
Replace value of header `X-Okapi-Tenant` in outgoing POST-requests to `/circulation/request` with ID of target tenant.

#### TODOS and Open Questions
- [ ] Use GitHub checklists. When solved, check the box and explain the answer.

## Learning
_Describe the research stage. Add links to blog posts, patterns, libraries or addons used to solve this problem._

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -71,9 +73,22 @@
...
+        "value": "http://okapi:9130"
+      },
+      {
+        "name": "KAFKA_HOST",
+        "value": "kafka"
+      },
+      {
+        "name": "KAFKA_PORT",
+        "value": "9092"
+      },
```
Comment (OleksandrVidinieiev) [inline descriptors/ModuleDescriptor-template.json:91]: This has nothing to do with this PR. Forgot to add these variables in my previous PR #7

File: src/main/java/org/folio/client/CirculationClient.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import org.folio.domain.dto.Request;
+import org.folio.spring.config.FeignClientConfiguration;
+import org.springframework.cloud.openfeign.FeignClient;
+import org.springframework.web.bind.annotation.PostMapping;
+
+@FeignClient(name = "circulation", url = "${folio.okapi-url}", configuration = FeignClientConfiguration.class)
+public interface CirculationClient {
+
+  @PostMapping("/circulation/requests")
+  Request createRequest(Request request);
```
Comment (alexanderkurash) [inline src/main/java/org/folio/client/CirculationClient.java:12]: createRemoteRequest?

File: src/test/java/org/folio/controller/KafkaEventListenerTest.java [hunk 3]
```diff
@@ -93,7 +93,7 @@ private static void publishEvent(String topic, String payload) {
 
   private static void waitForOffset(String topic, String consumerGroupId, int expectedOffset) {
     Awaitility.await()
-      .atMost(30, TimeUnit.SECONDS)
+      .atMost(60, TimeUnit.SECONDS)
```
Comment (alexanderkurash) [inline src/test/java/org/folio/controller/KafkaEventListenerTest.java:96]: Is this really necessary?

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -43,6 +43,10 @@ folio:
       enabled: true
   environment: ${ENV:folio}
   okapi-url: ${OKAPI_URL:http://okapi:9130}
+  logging:
+    feign:
+      enabled: true
+      level: full
```
[the remark this answers is not in the document]
Comment (OleksandrVidinieiev) [reply src/main/resources/application.yml:49]: Yes, but only when DEBUG level is enabled

File: src/test/java/org/folio/controller/KafkaEventListenerTest.java (hunk 3 above)
Comment (OleksandrVidinieiev) [reply src/test/java/org/folio/controller/KafkaEventListenerTest.java:96]: Kafka event consumption test is still failing on Jenkins by timeout occasionally. Increasing this value seems to help a bit.

File: src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java [hunk 5]
```diff
@@ -31,8 +36,20 @@ public Optional<EcsTlr> get(UUID id) {
   @Override
   public EcsTlr post(EcsTlr ecsTlr) {
     log.debug("post:: parameters ecsTlr: {}", () -> ecsTlr);
+    createRequest(ecsTlr, "university"); // TODO: replace with real tenantId
 
     return requestsMapper.mapEntityToDto(ecsTlrRepository.save(
       requestsMapper.mapDtoToEntity(ecsTlr)));
   }
+
+  private Request createRequest(EcsTlr ecsTlr, String tenantId) {
```
[the remark this answers is not in the document]
Comment (OleksandrVidinieiev) [reply src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java]: renamed

**Review by alexanderkurash — APPROVED** [review]

File: src/main/resources/swagger.api/schemas/EcsTlr.yaml [hunk 6]
```diff
@@ -23,6 +23,10 @@ EcsTlr:
       description: "Date when the request expires"
       type: string
       format: date-time
+    requestDate:
```
Comment (roman-barannyk) [inline src/main/resources/swagger.api/schemas/EcsTlr.yaml:26]: this field is required in the request schema of mod-circulation-storage. Should we make it required here as well?

File: src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java [hunk 7]
```diff
@@ -31,8 +36,20 @@ public Optional<EcsTlr> get(UUID id) {
...
       requestsMapper.mapDtoToEntity(ecsTlr)));
   }
+
+  private Request createRemoteRequest(EcsTlr ecsTlr, String tenantId) {
+    log.info("createRemoteRequest:: creating remote request for ECS TLR {} and tenant {}", ecsTlr.getId(), tenantId);
+    Request mappedRequest = requestsMapper.mapDtoToRequest(ecsTlr);
+    Request createdRequest = tenantScopedExecutionService.execute(tenantId,
+      () -> circulationClient.createRequest(mappedRequest));
+    log.info("createRemoteRequest:: request created: {}", createdRequest.getId());
+    log.debug("createRemoteRequest:: request={}", () -> createdRequest);
```
Comment (roman-barannyk) [inline src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java:51]:
```suggestion
    log.debug("createRemoteRequest:: request: {}", () -> createdRequest);
```

**Review by roman-barannyk — APPROVED** [review]

**Pull request merged**